File: Runtime/Providers/Apple/Plugins/Android/AppleAuthWebViewActivity.java
package com.greeneyes.auth;

import android.app.Activity;
import android.net.Uri;
import android.os.Bundle;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.unity3d.player.UnityPlayer;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.Base64;

public class AppleAuthWebViewActivity extends Activity {

    public static final String KEY_AUTH_URL       = "GREENEYES_AUTH_URL";
    public static final String KEY_REDIRECT_URL   = "GREENEYES_REDIRECT_URL";
    public static final String KEY_CALLBACK_OBJECT = "GREENEYES_CALLBACK_OBJECT";

    private String mRedirectUrl;
    private String mCallbackObject;
    private boolean mResultSent = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mRedirectUrl    = getIntent().getStringExtra(KEY_REDIRECT_URL);
        mCallbackObject = getIntent().getStringExtra(KEY_CALLBACK_OBJECT);
        String authUrl  = getIntent().getStringExtra(KEY_AUTH_URL);

        WebView webView = new WebView(this);
        WebSettings settings = webView.getSettings();
        settings.setJavaScriptEnabled(true);
        settings.setDomStorageEnabled(true);

        webView.setWebViewClient(new WebViewClient() {
            @Override
            public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
                return handleUrl(request.getUrl().toString());
            }

            @Override
            @SuppressWarnings("deprecation")
            public boolean shouldOverrideUrlLoading(WebView view, String url) {
                return handleUrl(url);
            }

            @Override
            public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
                if (request.isForMainFrame()) {
                    sendFailure("network");
                }
            }
        });

        webView.loadUrl(authUrl);
        setContentView(webView);
    }

    @Override
    public void onBackPressed() {
        sendFailure("cancelled");
        super.onBackPressed();
    }

    // Returns true if the URL is our redirect and was handled
    private boolean handleUrl(String url) {
        if (mRedirectUrl == null || !url.startsWith(mRedirectUrl)) {
            return false;
        }

        // Apple returns tokens in the URL fragment: redirect_uri#code=...&id_token=...
        Uri uri = Uri.parse(url);
        String fragment = uri.getFragment();

        if (fragment == null || fragment.isEmpty()) {
            sendFailure("invalidResponse");
            finish();
            return true;
        }

        try {
            // Parse fragment as key=value pairs
            String code = null, idToken = null, email = null, fullName = null;

            for (String param : fragment.split("&")) {
                String[] kv = param.split("=", 2);
                if (kv.length < 2) continue;
                String key = kv[0];
                String value = URLDecoder.decode(kv[1], "UTF-8");

                switch (key) {
                    case "code":     code    = value; break;
                    case "id_token": idToken = value; break;
                    case "user":
                        // Only provided on first sign-in
                        String[] nameEmail = parseUserParam(value);
                        fullName = nameEmail[0];
                        email    = nameEmail[1];
                        break;
                }
            }

            if (idToken == null || code == null) {
                sendFailure("invalidResponse");
                finish();
                return true;
            }

            String userId = extractSubFromJwt(idToken);

            JSONObject payload = new JSONObject();
            payload.put("identityToken",    idToken);
            payload.put("authorizationCode", code);
            payload.put("userId",           userId);
            if (email    != null && !email.isEmpty())    payload.put("email",    email);
            if (fullName != null && !fullName.isEmpty()) payload.put("fullName", fullName);

            sendSuccess(payload.toString());

        } catch (Exception e) {
            sendFailure("invalidResponse");
        }

        finish();
        return true;
    }

    // Parses Apple's user JSON: {"name":{"firstName":"John","lastName":"Doe"},"email":"john@example.com"}
    // Returns [fullName, email]
    private String[] parseUserParam(String userJson) {
        String fullName = null, email = null;
        try {
            JSONObject user = new JSONObject(userJson);
            email = user.optString("email", null);

            JSONObject name = user.optJSONObject("name");
            if (name != null) {
                String first = name.optString("firstName", "");
                String last  = name.optString("lastName", "");
                fullName = (first + " " + last).trim();
                if (fullName.isEmpty()) fullName = null;
            }
        } catch (JSONException ignored) { }
        return new String[]{fullName, email};
    }

    // Decodes the JWT payload and extracts the "sub" claim as userId
    private String extractSubFromJwt(String jwt) {
        try {
            String[] parts = jwt.split("\\.");
            if (parts.length < 2) return "";

            String payloadBase64 = parts[1];
            byte[] decodedBytes = Base64.getUrlDecoder().decode(payloadBase64);
            String payload = new String(decodedBytes, StandardCharsets.UTF_8);

            return new JSONObject(payload).optString("sub", "");
        } catch (Exception e) {
            return "";
        }
    }

    private void sendSuccess(String json) {
        if (mResultSent) return;
        mResultSent = true;
        UnityPlayer.UnitySendMessage(mCallbackObject, "OnSuccess", json);
    }

    private void sendFailure(String errorCode) {
        if (mResultSent) return;
        mResultSent = true;
        UnityPlayer.UnitySendMessage(mCallbackObject, "OnFailure", errorCode);
    }
}

File: Runtime/Providers/Apple/Plugins/Android/AppleAuthWebViewActivity.kt
package com.greeneyes.auth

import android.app.Activity
import android.net.Uri
import android.os.Bundle
import android.view.KeyEvent
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import com.unity3d.player.UnityPlayer
import org.json.JSONException
import org.json.JSONObject
import java.net.URLDecoder
import java.util.Base64

class AppleAuthWebViewActivity : Activity() {

    companion object {
        const val KEY_AUTH_URL        = "GREENEYES_AUTH_URL"
        const val KEY_REDIRECT_URL    = "GREENEYES_REDIRECT_URL"
        const val KEY_CALLBACK_OBJECT = "GREENEYES_CALLBACK_OBJECT"
    }

    private var mRedirectUrl: String? = null
    private var mCallbackObject: String? = null
    private var mResultSent = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mRedirectUrl    = intent.getStringExtra(KEY_REDIRECT_URL)
        mCallbackObject = intent.getStringExtra(KEY_CALLBACK_OBJECT)
        val authUrl     = intent.getStringExtra(KEY_AUTH_URL)

        val webView = WebView(this)
        webView.settings.apply {
            javaScriptEnabled = true
            domStorageEnabled = true
        }

        webView.webViewClient = object : WebViewClient() {
            override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
                return handleUrl(request.url.toString())
            }

            @Suppress("DEPRECATION")
            override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
                return handleUrl(url)
            }

            override fun onReceivedError(view: WebView, request: WebResourceRequest, error: WebResourceError) {
                if (request.isForMainFrame) sendFailure("network")
            }
        }

        webView.loadUrl(authUrl ?: "")
        setContentView(webView)
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            sendFailure("cancelled")
            return true
        }
        return super.onKeyDown(keyCode, event)
    }

    // Returns true if the URL is our redirect and was handled
    private fun handleUrl(url: String): Boolean {
        val redirectUrl = mRedirectUrl ?: return false
        if (!url.startsWith(redirectUrl)) return false

        // Apple returns tokens in the URL fragment: redirect_uri#code=...&id_token=...
        val fragment = Uri.parse(url).fragment
        if (fragment.isNullOrEmpty()) {
            sendFailure("invalidResponse")
            finish()
            return true
        }

        try {
            var code: String? = null
            var idToken: String? = null
            var email: String? = null
            var fullName: String? = null

            for (param in fragment.split("&")) {
                val kv = param.split("=", limit = 2)
                if (kv.size < 2) continue
                val value = URLDecoder.decode(kv[1], "UTF-8")

                when (kv[0]) {
                    "code"     -> code    = value
                    "id_token" -> idToken = value
                    "user"     -> {
                        // Only provided on first sign-in
                        val (parsedFullName, parsedEmail) = parseUserParam(value)
                        fullName = parsedFullName
                        email    = parsedEmail
                    }
                }
            }

            if (idToken == null || code == null) {
                sendFailure("invalidResponse")
                finish()
                return true
            }

            val payload = JSONObject().apply {
                put("identityToken",     idToken)
                put("authorizationCode", code)
                put("userId",            extractSubFromJwt(idToken))
                if (!email.isNullOrEmpty())    put("email",    email)
                if (!fullName.isNullOrEmpty()) put("fullName", fullName)
            }

            sendSuccess(payload.toString())

        } catch (e: Exception) {
            sendFailure("invalidResponse")
        }

        finish()
        return true
    }

    // Parses Apple's user JSON: {"name":{"firstName":"John","lastName":"Doe"},"email":"john@example.com"}
    // Returns Pair(fullName, email)
    private fun parseUserParam(userJson: String): Pair<String?, String?> {
        return try {
            val user  = JSONObject(userJson)
            val email = user.optString("email").ifEmpty { null }

            val name     = user.optJSONObject("name")
            val fullName = name?.let {
                "${it.optString("firstName", "")} ${it.optString("lastName", "")}".trim().ifEmpty { null }
            }

            Pair(fullName, email)
        } catch (ignored: JSONException) {
            Pair(null, null)
        }
    }

    // Decodes the JWT payload and extracts the "sub" claim as userId
    private fun extractSubFromJwt(jwt: String): String {
        return try {
            val parts = jwt.split(".")
            if (parts.size < 2) return ""

            val decoded = Base64.getUrlDecoder().decode(parts[1])
            JSONObject(String(decoded, Charsets.UTF_8)).optString("sub", "")
        } catch (e: Exception) {
            ""
        }
    }

    private fun sendSuccess(json: String) {
        if (mResultSent) return
        mResultSent = true
        UnityPlayer.UnitySendMessage(mCallbackObject, "OnSuccess", json)
    }

    private fun sendFailure(errorCode: String) {
        if (mResultSent) return
        mResultSent = true
        UnityPlayer.UnitySendMessage(mCallbackObject, "OnFailure", errorCode)
    }
}
